dev/io.openliberty.microprofile.telemetry.1.1.internal/src/io/openliberty/microprofile/telemetry11/internal/rest/TelemetryServletFilter.java

```diff
@@ -0,0 +1,311 @@
+            Tr.debug(tc, "isAsyncStarted=" + request.isAsyncStarted());
+        }
+
+        if (request.isAsyncStarted()) {
+
+            AsyncContext asyncContext = request.getAsyncContext();
+            asyncContext.addListener(new AsyncListener() {
+                @Override
+                public void onComplete(AsyncEvent event) throws IOException {
+                    endSpan(request, response, null);
+                }
+
+                @Override
+                public void onTimeout(AsyncEvent event) throws IOException {
+                    endSpan(request, response, event.getThrowable());
+                }
+
+                @Override
+                public void onError(AsyncEvent event) throws IOException {
+                    endSpan(request, response, event.getThrowable());
+                }
+
+                @Override
+                public void onStartAsync(AsyncEvent event) throws IOException {
+                }
+            });
+        } else {
+            endSpan(request, response, null);
+        }
+
+        if (scope != null) {
+            scope.close();
+            request.removeAttribute(SPAN_SCOPE);
+        }
+
+    }
+
+    private void endSpan(ServletRequest request, ServletResponse response, Throwable throwable) {
+        if (instrumenter != null) {
+            Context spanContext = (Context) request.getAttribute(SPAN_CONTEXT);
+            if (spanContext == null) {
+                return;
+            }
+
+            try {
+                if (tc.isDebugEnabled()) {
+                    Tr.debug(tc, "End span traceId=" + Span.fromContext(spanContext).getSpanContext().getTraceId() + " spanId="
+                                 + Span.fromContext(spanContext).getSpanContext().getSpanId());
+                }
+                instrumenter.end(spanContext, request, response, throwable);
+            } finally {
+                request.removeAttribute(SPAN_CONTEXT);
+                request.removeAttribute(SPAN_PARENT_CONTEXT);
+            }
+        }
+    }
+
+    private static class ServletRequestContextTextMapGetter implements TextMapGetter<ServletRequest> {
+
+        @Override
+        public Iterable<String> keys(final ServletRequest request) {
+            if (request instanceof HttpServletRequest) {
+                HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+                return new HashSet<String>(Collections.list(httpServletRequest.getHeaderNames()));
+            }
+            return Collections.emptyList();
+        }
+
+        @Override
+        public String get(final ServletRequest request, final String key) {
+            if (request == null) {
+                return null;
+            }
+
+            if (request instanceof HttpServletRequest) {
+                HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+                return httpServletRequest.getHeader(key);
+            }
+            return null;
```

Comment:
Wouldn't say no to some logging or guards, but I'm getting this is known to be safe by now.